feat(mesh): error status page

LinHeMa opened the pull request:
# Error Handling and UI Improvements

## Features

### Global Error Handling
- Implemented `global-error.tsx` to catch errors at the root level
- Added temporary hard-coded status codes (to be refactored later)

### 404 Not Found Handler
- Created `not-found.tsx` to handle invalid URLs

### Profile-specific Error Handling
- Added `profile/[member, publisher]/error.tsx` to manage errors within `profile/[member, publisher]/{id}`
- Maintained consistent layout with `profile/` section

## UI Enhancements
- Added SVG illustrations for 404 and 500 error pages
- Introduced new icon type for improved visual consistency
- Created global `error-page` component for standardized error display
- Implemented `primary-outlined` button style for enhanced UI options

## Technical Details
- File changes:
  - `global-error.tsx`
  - `not-found.tsx`
  - `profile/[member, publisher]/error.tsx`
  - Various UI component files (for SVG, icons, and button styles)

## Next Steps
- Refactor hard-coded status codes in `global-error.tsx`
- Consider expanding error handling to other sections of the application

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207975510192869

LinHeMa committed:
feat(error-status-page):init commit

LinHeMa committed:
feat(error-status): add error page to local and profile

erase2004 reviewed (commented):
1. `not-found.tsx` 不是只有 root 才可以有，其他路徑底下也可以有 [參考](https://nextjs.org/docs/app/building-your-application/routing#component-hierarchy)
2. 目前有些資料是否存在的檢查都放在 client side 去抓取，導致頁面剛開始 render 時，回傳的 http status code 是 200，應該把這部份的檢查移到 server side，並在資料不存在的情況下呼叫 `notFound` 方法，跳轉到對應的 `not-found.tsx` 呈現，同時讓回傳的 http status code 是 404
3. not-found.tsx 和 error.tsx 應該各司其職，不要在 error 裡面處理 not-found
4. 顯示範圍超過主內容的右邊界
![Snipaste_2024-08-27_00-15-02](https://github.com/user-attachments/assets/55b469b4-55d3-427b-ab1a-8a01b9664cd4)

tyc67 commented on packages/mesh/app/global-error.tsx:
```diff
@@ -0,0 +1,23 @@
+  error: Error & { digest?: string }
+  reset: () => void
+}) {
+  const statusCode = error.digest?.includes('NEXT_NOT_FOUND') ? 404 : 500
```
`error.digest` 應該是一個 hash [doc](https://nextjs.org/docs/app/api-reference/file-conventions/error#errordigest) , 這邊可能不會如預期的 conditional 判斷

LinHeMa replied:
了解，這裡在[d8769f2](https://github.com/readr-media/Sachiel/pull/834/commits/d8769f2af0451c94130258f822abde20cdedf0cd)把`not-found`與`error`分開，不使用`digest`進行判斷

erase2004 replied:
statusCode 應為 500，這一行可以移除

LinHeMa committed:
fix(error-status): separate not-found and error

LinHeMa committed:
fix(error-status): max content width

LinHeMa commented:
- [d8769f2](https://github.com/readr-media/Sachiel/pull/834/commits/d8769f2af0451c94130258f822abde20cdedf0cd) 處理了profile資料夾內`not-found`與`error`的分開、並且在server action如果有錯誤透過`next/navigation`的`notFound`進行錯誤處理
- [5baab1e](https://github.com/readr-media/Sachiel/pull/834/commits/5baab1e0e84dba6207be06d7588ae902c7b16dec) 處理了範圍超過主內容右邊界的問題

erase2004 commented on packages/mesh/app/actions/get-profile.ts:
在 try catch block 中呼叫 `notFound` 是有問題，請[參考](https://nextjs.org/docs/app/api-reference/functions/unstable_rethrow) (沒有要用這個 function

erase2004 replied:
你可以直接在 try catch 之前直接呼叫 `notFound` 來觀察，由於使用的是在 client side (`useProfileState`)，而非 server component，導致不會跳轉到 notFound 樣式

LinHeMa replied:
[c3ff9b5](https://github.com/readr-media/Sachiel/pull/834/commits/c3ff9b5bfd18a720a0dc151ce7c1a57d77c752b9) 進行處理

erase2004 commented on packages/mesh/app/profile/publisher/[publisherId]/page.tsx:
這是一個正確的改動，在 Server Component 中呼叫 notFound，可以讓 NotFoundBoundary catch 到，並顯示 not-found 的樣式，其他有呈現 not-found 需求的路徑，應在 Server Component 上去檢查資料是否不存。

erase2004 commented on packages/mesh/app/global-error.tsx:
```diff
@@ -0,0 +1,23 @@
+  const statusCode = error.digest?.includes('NEXT_NOT_FOUND') ? 404 : 500
+  // TODO: cannot add stateful header
+  return (
+    <html>
```
注意到少了 lang 和字型設定，應與 layout.tsx 同步

LinHeMa replied:
[b3dac0f](https://github.com/readr-media/Sachiel/pull/834/commits/b3dac0f7601c462ba8b4464bfa7e737f8090154b) 進行處理

erase2004 commented on packages/mesh/app/not-found.tsx:
```diff
@@ -0,0 +1,26 @@
+export default function NotFound() {
+  const router = useRouter()
+  const handleRefresh = () => {
+    router.refresh()
```
`router.refresh()` 雖然會重拿 RSC payload，但與 `windows.location.reload` 整頁重整不同，
可能要釐清這部份的行為，預期會有怎樣的效果。
[參考](https://nextjs.org/docs/app/api-reference/functions/use-router#userouter)

LinHeMa replied:
[3dc1821](https://github.com/readr-media/Sachiel/pull/834/commits/3dc1821786d5a557be1a44c37a61e087968b0521) 進行處理

erase2004 reviewed (commented):
## 按鈕的 hover/active 效果似乎與設計不符？
https://github.com/user-attachments/assets/a9ecfd3d-2e0b-435e-928f-6d7d1f3c7912

## 如果 not-found 的整頁樣式，在不同路徑下都相同，那可以只用 `/app/not-found` 統一處理

LinHeMa committed:
fix(error-status): remove error digest

LinHeMa committed:
fix(error-status): keep global-error attr same as layout

LinHeMa committed:
fix(error-status): use window reload

LinHeMa committed:
Merge branch 'dev' of https://github.com/readr-media/Sachiel into feat/error-status-page

LinHeMa committed:
fix(error-status): remove duplicate

LinHeMa commented:
[0c4614e](https://github.com/readr-media/Sachiel/pull/834/commits/0c4614e1ff93fdb2148a953a4c162b29e6603426) 刪除`profile/member`, `profile/publisher`底下的`not-found.tsx`因為樣式重複

LinHeMa committed:
fix(error-status): change not found handle for client

erase2004 reviewed (commented):
現在載入 `/profile/member/[customId]` 會出現，先跑出 404 後再顯示內容情況。

https://github.com/user-attachments/assets/203fc22f-86a6-47e1-8690-3a8c59bdfa6c

erase2004 commented:
> ## 按鈕的 hover/active 效果似乎與設計不符？
> firefox_wucI8QTaWg.mp4

這個問題似乎還沒解決

LinHeMa committed:
feat(components): add primary-outlined-without-hover

LinHeMa committed:
feat(error-status): add profile/member error page

LinHeMa commented:
在[20244a9](https://github.com/readr-media/Sachiel/pull/834/commits/20244a963db3b6a956f943c2a28b4e9495122a67)修復button樣式問題

LinHeMa commented:
> 現在載入 /profile/member/[customId] 會出現，先跑出 404 後再顯示內容情況。

這個部分，merge @tyc67 的PR 本地測已經解決

tyc67 commented on packages/mesh/components/button.tsx:
```diff
@@ -13,6 +13,10 @@ const buttonColor = {
     'bg-blue-500 text-white disabled:bg-disable disabled:text-primary-400 disabled:cursor-not-allowed',
   primary:
     'bg-primary-700 text-white border border-transparent hover:bg-primary-800 disabled:bg-disable disabled:text-primary-400 disabled:cursor-not-allowed',
+  'primary-outlined':
```
這個 style 其他地方有用到嗎?

LinHeMa replied:
[1b868be](https://github.com/readr-media/Sachiel/pull/834/commits/1b868be3726181d869e8dbaaed4b35a0d1a0b977)刪除新增的style，改成直接修改既有的`primary-outlined`

LinHeMa committed:
style(button): edit primary outlined

erase2004 reviewed (commented):
1. 有 `/profile/error.tsx` 後，`/profile/member/error.tsx` 和 `/profile/publisher/error.tsx` 似乎就不必要了
2. 一樣有 profile 載入時，先出現 404 的問題
![vBEtTWHiwD](https://github.com/user-attachments/assets/b6d0be7b-39d4-462e-8614-00a37b8cf010)
3. error page 的按鈕有 hover/active 效果
https://www.figma.com/design/wqODc69zqgbu7TZFtQAmnN/2024_READr-Mesh_Web?node-id=9942-136662&node-type=SECTION&m=dev

LinHeMa committed:
Merge branch 'dev' of https://github.com/readr-media/Sachiel into feat/error-status-page

LinHeMa committed:
Merge branch 'dev' of https://github.com/readr-media/Sachiel into feat/error-status-page

LinHeMa committed:
feat(error-status): delete unnecessary error page

LinHeMa committed:
refactor(error-status): profile loading avoid 404

LinHeMa committed:
fix(error-status):

LinHeMa committed:
refactor(error-status): add btn hover/active

LinHeMa commented:
[86a50c6](https://github.com/readr-media/Sachiel/pull/834/commits/86a50c69e950c1071a3c50aedf89c8b8b0a08b3a)移除/profile/member/error.tsx 和 /profile/publisher/error.tsx

[d6f762f](https://github.com/readr-media/Sachiel/pull/834/commits/d6f762f4e3a2430d264045c1bb72b9d9451f261e)[d6f762f](https://github.com/readr-media/Sachiel/pull/834/commits/d6f762f4e3a2430d264045c1bb72b9d9451f261e) 解決載入先出現404問題

[65900d8](https://github.com/readr-media/Sachiel/pull/834/commits/65900d835bcbe975b5dafbcd1a3b1e2f8149d3a1) btn hover, active 樣式

LinHeMa committed:
Merge branch 'dev' of https://github.com/readr-media/Sachiel into feat/error-status-page

erase2004 reviewed (commented):
* Spinner 的顯示範圍不正確
![Snipaste_2024-09-10_13-39-42](https://github.com/user-attachments/assets/371b262b-3fb5-46e2-add5-5e9194d6e543)
* profile 頁面的呈現，會是 無資料 -> Spinner -> 有資料 / 404，應該調整為 Spinner -> 有資料 / 404 ?
![firefox_slcm8NADaa](https://github.com/user-attachments/assets/2d1ff858-cbf1-46cb-8b00-c8f8870e2fe0)

LinHeMa committed:
Merge branch 'feat/error-status-page' of github.com:LinHeMa/Sachiel into feat/error-status-page

LinHeMa committed:
refactor(error-status): render order and hook

LinHeMa commented:
[b2ec5ba](https://github.com/readr-media/Sachiel/pull/834/commits/b2ec5ba4f1ea59d75139f503c69a5b29f823c656) 將`loading`與`error`判斷提早，避免無資料先出現的問題。

erase2004 commented on packages/mesh/app/profile/member/[customId]/_components/profile-page.tsx:
```diff
@@ -83,53 +105,37 @@ const ProfilePage: React.FC<ProfilePageProps> = ({ isMember }) => {
-        name={name}
-      />
-    </>
+    <Suspense fallback={<Spinner />}>
```
Suspense 是多餘的，這種在 client side 抓取資料的情境，不適用於 suspense，參閱 [Note](https://react.dev/reference/react/Suspense#displaying-a-fallback-while-content-is-loading)

LinHeMa replied:
透過[764aa40](https://github.com/readr-media/Sachiel/pull/834/commits/764aa40103a8d47a35740371e764025262f9c0b0)處理，
並且將spinner對齊header最大寬度

erase2004 reviewed (commented):
Spinner 的顯示範圍不正確 的問題未被解決

LinHeMa committed:
refactor(error-status): spinner location and suspense

erase2004 reviewed (commented):
最大寬度不正確
![Snipaste_2024-09-11_09-40-42](https://github.com/user-attachments/assets/a2d9eb08-abfd-4901-a193-1e97f4b20459)

LinHeMa committed:
style(error-status): max width

- change all max width to tailwind config variable.
- change loading view to correct max width.

LinHeMa commented:
[7a14265](https://github.com/readr-media/Sachiel/pull/834/commits/7a14265082bff33a4eec67e5856715a3a74e4f21) 處理最大寬度的問題，同時將個人檔案頁面的`1120px`都替換成tailwind config變數

erase2004 reviewed (approved):
感謝一併調整 👏
